### PR TITLE
Add implicit roles to td, th, and simplify tr

### LINKED ISSUE #195: Make gridcell one of the implicit roles for td elements (if in grid)
Reopening #60 because ARIA in HTML is now inconsistent with HTML AAM for "gridcell", and because browsers are now using implied role of "gridcell" when `td` is in "grid".

The HTML AAM spec says that in certain circumstances, td (and th) map to gridcell:
- [td (ancestor table element has grid role)](https://w3c.github.io/html-aam/#el-td-gridcell)
- [th (is neither column header nor row header, and ancestor table element has grid role)](https://w3c.github.io/html-aam/#el-th-gridcell)

To see this in Chrome:
1. [APG Data Grid Example](http://w3c.github.io/aria-practices/examples/grid/dataGrids.html)
2. inspect one of the grid's cells in Chrome DevTools
3. the td element in the Accessibility panel has (implicit) role "gridcell"

image

Firefox also defaults to `role="gridcell"` for `td` in `grid`. Did not check Safari, but Chrome + FF = 2.

So the [ARIA in HTML entry for td](http://w3c.github.io/html-aria/#el-td) (and th) need to say basically the same thing as HTML AAM for td and th.

Also, the [tr entry](http://w3c.github.io/html-aria/#el-tr) seems [a bit oddly worded](https://github.com/w3c/html-aria/issues/62) now, given that `tr` in `grid` (and `treegrid`) defaults to `row`.

It currently says:
> role=row, may be explicitly declared when child of a table element with role=grid or treegrid

but I think it only needs to say:
> role=row, when child of a table element with role=grid or treegrid

I will provide a PR shortly.

### PR DESCRIPTION
**Pull request by carmacleod**
Resolves #195.
- add gridcell to td
- add cell and gridcell to th
- simplify tr

***
<a href="https://pr-preview.s3.amazonaws.com/carmacleod/html-aria/pull/196.html" title="Last updated on Aug 13, 2020, 2:59 PM UTC (1058ce8)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/html-aria/196/a7bf6c0...carmacleod:1058ce8.html" title="Last updated on Aug 13, 2020, 2:59 PM UTC (1058ce8)">Diff</a>

### CODE REVIEW

**Comment by carmacleod** [comment]
Not sure what remains to be done here?
Perhaps one or both of the thoughts in https://github.com/w3c/html-aria/issues/195#issuecomment-561184189 ?

File: index.html [hunk 1]
```diff
@@ -2271,8 +2274,14 @@ <h2 id="docconformance">
               [^th^]
             </td>
             <td>
-              <code>role=<a href="#index-aria-columnheader">columnheader</a></code> or
-              <a href="#index-aria-rowheader">`rowheader`</a>
+              <p>
+                If a descendant of a `table` element, <code>role=<a href="#index-aria-columnheader">columnheader</a></code> or
+                <a href="#index-aria-rowheader">`rowheader`</a> or <a href="#index-aria-rowheader">`cell`</a> (if not a header).
```
Comment (scottaohara) [inline index.html:2656]: we need to verify this, specifically if browsers will expose a `th` as a `cell` under different conditions, or if it's consistently exposed as a header.

if the former, then this is good.  if the latter, then we need to determine if that's appropriate or not.  and if not, then this becomes an issue for updating HTML AAM as well.

File: index.html [hunk 2]
```diff
@@ -2228,7 +2228,10 @@ <h2 id="docconformance">
                 <code>role=<a href="#index-aria-cell">cell</a></code> if a
                 descendant of a `table` element.
               </p>
-              <!--<p><code>role=<a href="#index-aria-cell">gridcell</a></code> if a descendant of a <code>table</code> elementwith <code>role=grid</code></p>-->
+              <p>
+                <code>role=<a href="#index-aria-gridcell">gridcell</a></code> if a
```
Comment (scottaohara) [inline index.html:2632]: we want to verify this actually is true in browsers.  it _should_ be, but if not, then we'd want to note a special case for it and file bugs where this doesn't hold true.

File: index.html [hunk 3]
```diff
@@ -2290,9 +2299,7 @@ <h2 id="docconformance">
               [^tr^]
             </td>
             <td>
-              <a href="#index-aria-row">`role=row`</a>, may be
-              explicitly declared when child of a `table` element
-              with `role=grid` or `treegrid`
+              <a href="#index-aria-row">`role=row`</a>
```
Comment (scottaohara) [inline index.html:2681]: this *should* be what we need here.  again, we want to just make sure that there aren't instances where this is not exposed correctly, so as to not remove a special case for why authors would need to assign the implicit role.

**Comment by stevefaulkner** [comment]
@carmacleod @scottaohara what is needed to move this along? testing? Should this not be surfaced in HTML API spec first?

**Comment by carmacleod** [comment]
@stevefaulkner I guess @scottaohara asked for more testing.
So here's a bit more, using [the grid/table codepen](https://codepen.io/stevef/pen/OmVeJL) from #60 (with added treegrid):

#### Windows Chrome:
- `tr` in `table/grid/treegrid` is `row` ✔️ 
- `th` in `table/grid/treegrid` is either `columnheader` or `rowheader` depending on AllTheThings ✔️ 
- `td` in `grid/treegrid` is `gridcell` ✔️ 
- **Chrome bug:** `td` in **`table`** is `gridcell` but should be `cell` ❌ 
(this must just be a typo in the DevTools a11y tree, because AT all work with tables... either that, or AT are ignoring `gridcell` and assuming `cell` in a `table`).

Screenshot of Chrome Accessibility tree with **grid** thead row selected:

image

Screenshot of Chrome Accessibility tree with **grid** tbody row selected:

image

Screenshot of Chrome Accessibility tree with **table** tbody row selected:

image

#### Windows Firefox:
- **FF bug:** Accessibility tree says role=table even though role="grid" is explicitly set ❌ 
which may be because [the code for table and grid hasn't been split yet](https://bugzilla.mozilla.org/show_bug.cgi?id=1177270)?
(or maybe that's just a typo in the FF DevTools? Child td's know they're in a grid).
- Accessibility tree says role="tree table" when role=treegrid ❓ 
(is that correct? name changes in DevTools are not helpful - devs are expecting the role names from the specs)
- `tr` in `table/grid/treegrid` is `row` ✔️ 
- `th` in `table/grid/treegrid` is either `columnheader` or `rowheader` depending on AllTheThings ✔️ 
- `td` in `table` is `cell` and in `grid` is `gridcell` ✔️ 
- **FF bug:** `td` in `treegrid` is `cell` ❌ 

Screenshot of FF Accessibility tree for `table role=grid` (which FF incorrectly calls "table"):

image

Screenshot of FF Accessibility tree for `table role=treegrid` (which FF confusingly calls "tree table"):

image

#### macOS Safari:
- `tr` in `table/grid/treegrid` is `row` ✔️ 
- `th` in `table/grid/treegrid` is either `columnheader` or `rowheader` depending on AllTheThings ✔️ 
- `td` in `table` is `cell` ✔️ 
- **Bug in Safari:** `td` in `grid/treegrid` is `cell` (should be `gridcell`) ❌ 

Screenshot of Safari Accessibility tree with grid thead row selected:

image

Screenshot of Safari Accessibility tree with grid tbody row selected:

image

So, other than some naming weirdness in the DevTools a11y trees, the only "real" issues have to do with `td` defaulting to the wrong role.

#### Questions:
1. From @stevefaulkner:
> Should this not be surfaced in HTML API spec first?

Not completely sure if "HTML API spec" means "HTML AAM"? If yes, then please note that HTML AAM spec has already surfaced td and th mappings to gridcell when in grid, as follows:
- [td (ancestor table element has grid role)](https://w3c.github.io/html-aam/#el-td-gridcell)
- [th (is neither column header nor row header, and ancestor table element has grid role)](https://w3c.github.io/html-aam/#el-th-gridcell)

Unfortunately, HTML AAM doesn't mention `treegrid` for `th` or `td` - shouldn't it? ([treegrid was in ARIA 1.0](https://www.w3.org/TR/wai-aria-1.0/roles#treegrid) but I don't know when it was first supported in UA/AT)

2. From @scottaohara in https://github.com/w3c/html-aria/issues/195#issuecomment-561325871:
> `th`s could be exposed as a `cell`... though testing that again today with modern browsers, I'm not sure that's actually still the case... 

How does one mark up a `table` so that a `th` is heuristically determined to be a `cell`? (i.e. HTML AAM says, "th is neither column header nor row header")

3. Also, @scottaohara mentioned in https://github.com/w3c/html-aria/issues/195#issuecomment-560693403 that "Any role" on tr, th, td was a bit too permissive.

So we need to at least make authors think about the parent table's role. Some possible words:
1. > Any `role` applicable to the ancestor `table` element's allowed roles and implied role (if any).
2. > Any `role` applicable to the ancestor `table` element's explicit or implied role.
3. > No `role` if the ancestor `table` element has role=table, grid, or treegrid; otherwise any `role`.

Please let me know if any of those 3 phrases makes sense for tr, th, and td for their "ARIA roles, states and properties which MAY be used" column, or if something else is preferred.

**Comment by stevefaulkner** [comment]
@scottaohara Build failed?

**Comment by patrickhlauke** [comment]
seems related to respec validation https://travis-ci.com/github/w3c/html-aria/builds/177415969

**Comment by carmacleod** [comment]
Submitted a few issues:
1. Chrome: [td in table is gridcell but should be cell](https://bugs.chromium.org/p/chromium/issues/detail?id=1111353)
2. Firefox: [Accessibility tree says role=table even though role="grid" is explicitly set](https://bugzilla.mozilla.org/show_bug.cgi?id=1656330)
3. Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1656339

Still need to submit 1 or 2 more...
Edit: Opened https://github.com/w3c/html-aam/issues/298 and submitted a pr and it's merged already - yay! :)

**Comment by scottaohara** [comment]
@carmacleod ok.  finally back to this. last thing i think we need here is the update to the "Any role" for th, tr, td.

I like option 3 "No role if the ancestor table element has role=table, grid, or treegrid; otherwise any role."

**Comment by carmacleod** [comment]
@scottaohara 

> last thing i think we need here is the update to the "Any role" for th, tr, td.

Done

**Review by scottaohara — APPROVED** [review]

**Pull request merged**